### ISSUE TITLE
Battle Calculator and Dice Roll Menus

### ISSUE THREAD

**Issue by ACG8** [sIKE23/Mage-Wars#144]
I have programmed the rudiments of what will be code that tailors an attack roll to the creature attacking and the one defending. Now, it only remains to improve upon it. The commit that defines the basics is here: https://github.com/sIKE23/Mage-Wars/commit/34f2c64256672127110c8e499c8f8974d2b30dd0

Here are some immediate things that I want to work on with this:
- Allow similar functionality when targeting a creature with an attack.
- Expand to include mage attacks
- State whether the results of the attack were above or below average (especially handy when deciding whether to use Akiro's Favor/Temple of the Dawnbreaker)
- Give players the ability to auto-apply damage to their own creatures on a case-by-case basis (compute the damage/effects resulting from the attack and then bring up a dialog box to ask the recipient whether they accept the calculated damage/effects).

In general, I want to move toward having the engine be much more aware of what is actually going on in the game, and this is the most obvious place to start.

**Comment by anandaguneratne** [comment]
Oh, and I completely forgot one of the really big improvements:
- Detect bonuses to creature when computing its attacks, including bonuses from attached enchantments.

**Comment by anandaguneratne** [comment]
https://github.com/sIKE23/Mage-Wars/commit/75c1b64c70182190341ff44a8706b6668fc4b393 consolidates all the dice menus into a single function.

**Comment by sIKE23** [comment]
Why aren't you using:

<event name="OnTargetCardArrow">     <param name="player" type="Player"/>
    <param name="fromCard" type="Card"/>     <param name="toCard" type="Card
"/>          <param name="
isTargeted" type="bool"/> </event>

**Comment by anandaguneratne** [comment]
Oh...That is a REALLY good idea...

**Comment by anandaguneratne** [comment]
https://github.com/sIKE23/Mage-Wars/commit/6c3aeabc5f475ea39b97869460a3ed3f3d6f5853 provides the first feature: support for attack spells.

**Comment by sIKE23** [comment]
Wouldn't you, when the event is triggered, then check to see if <TAB> has
been pressed?

**Comment by anandaguneratne** [comment]
No, I think the onTargetCardArrow just means when you draw an arrow it is triggered. No checking for TAB; that is a separate trigger.

**Comment by anandaguneratne** [comment]
Mages are integrated now, and attacks from equipment! See https://github.com/sIKE23/Mage-Wars/commit/97503dd8b083661443314e57634bbc518e762143. Also, the attackTarget function announces how much damage it thinks the attack did, and whether this amount is above or below average for that attack (to help with Temple of the Dawnbreaker/Akiro's Favor decisions).

**Comment by sIKE23** [comment]
This is some really slick work! I like the Tab key.......

**Comment by anandaguneratne** [comment]
The attack buttons now display information on attack traits and effects (and the probability of each effect).

**Comment by anandaguneratne** [comment]
Glad you like the TAB key. https://github.com/sIKE23/Mage-Wars/commit/63276a86f5c202a83266b18125cab0bbc8e15912 implements TAB for all attack rolls, now.

**Comment by anandaguneratne** [comment]
With the most recent commit, the basic functionality of the attack detection module (That isn't a great name; let's call it "battle calculator module" from now on) is complete; it works, and has pretty much all of the features I originally wanted. Now all that remains is to clean it up, and take care of the numerous special cases (also, to improve the detection of attacks available to mages and to improve detection of traits from enchantments/equipment, though the latter should be its own issue).

**Comment by sIKE23** [comment]
We need to make sure that players (rollDice -> genericAttack) can turnoff the Battle Calculator and the game basically from an attack perspective works as it does today, as there maybe people who do not prefer this level of automation. 

Also if one player turns it off what happens to the other player? This is why we have all of those questions at the start of the game, it is real hard to make assumptions for players that are not us :)

**Comment by sIKE23** [comment]
If I don't say it enough, amazing work.....

**Comment by anandaguneratne** [comment]
Thanks! I'll see what I can do about adding an off switch. In terms of declaring an attack, it should be pretty simple (just go to the generic attack menu); the tricky bit is when receiving an attack. I think probably the best option is to halt the attack process after declaring the attack, so that the player with automation may select the attack from a list, but after the dice are rolled nothing else happens (it would be up to the un-automated player to apply effects, or what-have-you).

You do bring up a good point - attack automation relies on both parties having it enabled. I would think that in most cases either both would enable or both would disable, but there could be cases where one enables and one disables, in which case it is essentially disabled for both, for all intents and purposes...

You know what? Maybe that is the solution. Instead of figuring out what happens if only one player has enabled the option, why don't we just disable it if at least one player has it disabled? So everybody has to agree to use it for it to be enabled. Sound like a good solution?

**Comment by anandaguneratne** [comment]
I accidentally closed this (without noticing).

**Comment by sIKE23** [comment]
The Gargoyle Sentry is not being calculated correctly when he is on Guard. What is interesting here is that he has a pre-req (Guarding) before he gets the Armor +3 and Tough -3 buffs. Do you think it is worth creating a prereq tag that you could write code to check for. 

i.e.

[prereq=Guard] [Armor +3] [Tough -3]

The BC would then know to check to see if the creature with this tag has a Guard Marker and if so then apply the buffs. Otherwise we (you) will have to write custom code each time we see something like this.

**Comment by ACG8** [comment]
No, I don't think tags are the best solution for this problem. I'll take a look at the Gargoyle's code (I thought I had it working; is it double-counting the bonus?) Long term (by which I mean Q2), I think we will be forced to create creature dictionaries in order to maintain the code (this is all right though; it will give us a lot of flexibility in implementing other features and will make things easier in the long run).

Exceptions like the Gargoyle aren't really much of an issue because they don't show up often. Writing a general purpose detector to take care of all the side cases would likely end up requiring the same amount of code as just coding the exceptions.

One problem with tags (though minor) is that they show up in the proxy generator.

**Comment by ACG8** [comment]
Okay, I found the problem. The exception was referring to 'Sentry Gargoyle' instead of 'Gargoyle Sentry'. Problem is now fixed.

**Comment by sIKE23** [comment]
We could create a property for items like this that is hidden from the
Deckbuilder and since that property is not  defined in the proxygen it will
not be sceen...

https://github.com/kellyelton/OCTGN/wiki/definition.xml#property

As for the Gargoyle his additional Armor is no being calculated when he is
Guarding. I haven't even tested the Tough -3

//FC

**Comment by sIKE23** [comment]
Do I need to remove the updated tags in the set.xml....

**Comment by ACG8** [comment]
Don't worry (remove it if you like), it doesn't matter for now since it is in the card text (the BC won't see it, since it is a creature). 

We could certainly modify the proxygen, but I don't think it will be necessary, because there are numerous benefits to making a spell dictionary anyway. I'm going to open an issue for that, and will outline my plan there.

**Comment by sIKE23** [comment]
This is like what Brine did in Magic?

**Comment by ACG8** [comment]
As I understand it, yes. Though possibly more ambitious.

**Comment by sIKE23** [comment]
I do not know if this is important. Take 2 creatures with melee only attacks, place each in a different zone. I have the BC off. Press Tab, you will get no legal attacks. Now press the Tab button again. You will get a roll dice window.

If you do it with the Targeting Recticle and the Tab from the Attacker the same thing will happen every time. 

for the Targeting Arrow and Reticle you should be able to do either:
```python
card.arrow(target, active = False)
```
or
```python
card.target(active = False)
```
once you are done to remove them....

**Comment by ACG8** [comment]
Okay, I just tried this out and did not get the error that you are reporting. If you have the BC off, pressing TAB should always bring up the generic menu, not the 'legal attacks' menu.

However, I did find that dragging an arrow still declares an attack using the BC even when the BC is switched off, so I'll need to fix that.

Regarding the targeting arrow/reticle: the arrow is removed when you declare an attack (but not if you cancel the attack, in case you want to drag an arrow for another reason). Do you think it should always be removed, even when you cancel the attack? I'll see about removing the reticle when an attack is declared.

**Comment by sIKE23** [comment]
Testing the Gargoyle again, and I am just verifying the calculation here:

![calc](https://cloud.githubusercontent.com/assets/5695175/6457460/b4ab4b58-c13b-11e4-8d4e-20fce69df780.jpg)
I have a Guard Marker on him so he has Tough -3

4-10 = 1 Burn; 11+ = 2 Burn

So the second part is right 0% chance for 2 Burns (effect roll of 12 - 3 = 9)

the question is the percentage on the 2nd part
shouldn't the chance be in the 40% range not the high 50% range (a effect roll of 7 -3 = 4) ?

My power of statistics are not the greatest it just seems a bit off to me...

**Comment by ACG8** [comment]
Tough -3 effectively shifts the range up by 3, so it becomes

7-13 = 1 Burn, 14+ = 2 burns

You can't roll a 13, so really that is 7-12 = 1 Burn

Which is...6 values out of 12. It should be 50%. I'll see what is wrong.

**Comment by ACG8** [comment]
Fixed it! The problem was that it was counting the full range (7-13) rather than truncating it at 12 (7-12). Should work correctly now.

**Comment by sIKE23** [comment]
Was testing this evening, Unicorn vs. Gargoyle (with Guard token). I moved the Unicorn into the zone with the Gargoyle and "attacked." BC is on and since it doesn't calculate for Charge I choose other Dice Amount, then choose 5 dice and nothing happens. The Dice do not roll, no error messages, like nothing happened.

Second issue much less pressing though, our favorite buddy the Gargoyle, when on Guard should have Armor +4 and the Expected Damage should be less than 1.5 correct? Tough +3 is not calculated either.
![new jpeg image](https://cloud.githubusercontent.com/assets/5695175/6681866/5fe98c40-cc3c-11e4-995d-a515effd883e.jpg)

**Comment by ACG8** [comment]
1. The BC does take charge into account. It just doesn't add the bonus until after you select the attack, since a prompt is needed to determine whether the creature moved. Do you think the prompt should be given when selecting the target, then?

2. Fixed the problem preventing dice from being rolled; will upload with next commit.

3. No, armor and guard are being computed correctly. Expected damage asymptotically approaches half the number of attack dice as armor goes to infinity (i.e. resilient). For 3 dice vs 4 armor, the expected damage is slightly above 1.5, but the BC rounds to the first decimal place because I thought that any more decimal places would not give useful information. I tested Tough -3, and it also works; without a Guard, Valshalla has a 33% chance of stunning the Gargoyle (9+); with it, this chance goes down to 8.3% (12+)

**Comment by sIKE23** [comment]
I had the Wizard Attack both the Gargoyle and Unicorn with an Arc Lightning
and got the same percentages. I will re-test and trying again.

**Comment by ACG8** [comment]
The same percentages for both the Daze and the Stun effects, or just for the Daze?

Generally, tough will reduce the probability of the higher effect (Stun), but will leave the lower one (Daze) unchanged.

**Comment by sIKE23** [comment]
I was testing late last night with out phase management (movement
calculations not working). Let me re-test to verify my findings.

Was just testing the Unicorn vs. Gargolye using the BC. I see what you did
and it does work, though I had no way to intuit that it was going to work
that way. I am not sure if you could move the Charge prompt to before the
BC prompt, it then should calculate the Average Damage for 5 dice against 4
Armor instead of the 3 dice against 4 armor.

During the Reset/Upkeep phase I am getting the following errors:
Inconsistent state. Marker not found on card.
Inconsistent state. Marker count invalid

After banging around with it for a bit I was not able to determine why. I
was playing with a Wizard (Air) and a Priest. THe Wizard only had out the
Gargoyle and the Priest the Unicorn. Whatever the issue is it is preventing
the BC from resetting movement and attacks.

**Comment by sIKE23** [comment]
How hard are you wanting to look at the Battle Calc for corner cases this
time around? Example, the Temple High Guard, he reorders when he can
counter strike and while on Guard he gets Melee +2. I know we are looking
at revamping how card works the next time around.

</sIKE>

**Comment by ACG8** [comment]
I think I am good on corner cases for Q1; I have (I think) all the mage abilities, and those are the main things. Not intending to implement temple high guard.

**Comment by sIKE23** [comment]
Testing with Schwenkgott today, we saw an issue with the Flaming Hellion it rolled 5 dice and it should of only rolled 4. I suspect that the Mages ring buffed the Hellions attack.....

![new jpeg image](https://cloud.githubusercontent.com/assets/5695175/6772493/cbb0d2f4-d0cd-11e4-97c1-7423ef823743.jpg)

**Comment by ACG8** [comment]
I can't quite read the cards attached to priest, but is one of them a curse? If so, the the Adra Warlock's Smoldering Curses ability would give the Priest the Flame +1 trait, which would explain the extra attack die.

**Comment by sIKE23** [comment]
The two Enchantments are Harmonize (testing the Channeling) and Bear
Strength. I will make sure next time to grab the game log.

Any more thoughts on the logging idea I proposed?

</sIKE>

**Comment by sIKE23** [comment]
I was able to reproduce the issue: 

![flaming hellion game test issue](https://cloud.githubusercontent.com/assets/5695175/6787905/d2e4622a-d163-11e4-8d0b-1b8130add0b1.jpg)

Game log is on the dev share under the testing folder or you can just go here:
https://www.dropbox.com/s/4bcl91k1y7fzxh5/Flaming%20Hellion%20Game%20Test%20Log.txt?dl=0

**Comment by ACG8** [comment]
You were right; the ring was the problem. Fixed now.

Regarding the logging idea: sounds good. Having a server to keep track of stats would be extremely useful.

**Comment by sIKE23** [comment]
Are these feasialbe?

def chatmsg(str):
    notify("{}".format(str))
    <add str to (what type of?) variable>

def privatechatmsg(str):
    whisper("{}".format(str))
    <add str to (what type of?) variable>

def debug(str):
    global debugMode
    if debugMode:
        whisper("Debug Msg: {}".format(str))
        <add str to (what type of?) variable>

the <add str to (what type of?) variable> is the crux of the idea...

**Comment by ACG8** [comment]
Okay. I see nothing wrong with that. As for what type of variable it should be, I think a list would be best, since it maintains the ordering and is easy to update. You could probably store it as a string in the OCTGN global variables (call it "GameLog" or something), and then export it as a .txt file upon request or when the game finishes. The name of the text file could be autogenerated, using the date and time to guarantee uniqueness.

for example:
```
def chatmsg(str):
    mute()
    GameLog = eval(getGlobalVariable('GameLog'))
    notify("{}".format(str))
    GameLog.append(str)
    setGlobalVariable('GameLog', str(GameLog))
```

When exporting it, you could use ```"\n".split(GameLog)``` to separate the entries of GameLog into nice orderly lines (actually, I will have to refresh my memory on how writing to txt files works; a list might actually not need any additional processing).

**Comment by sIKE23** [comment]
That makes a lot of since would we do the following for debug and private?

def privatechatmsg(str):
    mute()
    GameLog = eval(getGlobalVariable('GameLog'))
    whisper("{}".format(str))
    GameLog.append(str)
    setGlobalVariable('GameLog', str(GameLog))

def debug(str):
    mute()
    global debugMode
    if debugMode:
        GameLog = eval(getGlobalVariable('GameLog'))
        whisper("Debug Msg: {}".format(str))
        GameLog.append(str)
        setGlobalVariable('GameLog', str(GameLog))

**Comment by ACG8** [comment]
Yes, assuming we want both of those to appear in the final log.

**Comment by sIKE23** [comment]
Do you assume not? I think they should do you have any reason why we
shouldn't?

**Comment by ACG8** [comment]
No, I don't have any reason why we shouldn't.

**Comment by sIKE23** [comment]
Did almost an hour tonight putting the BC through the paces. I have to say very nice work. Dazes work perfectly. Though Stun markers still allow the creature to attack. This item can added as an Enhancement for the Q2 release if you think it is a good idea.

When a creature uses an attack using the Arrow how hard would it be to Exhaust his Action Marker. Much like the Stun item above a creature with a Exhausted Action Marker should not be able to Attack. Both of these items we can added as an Enhancement for the Q2 release if you think they are good ideas.

**Comment by ACG8** [comment]
Yes, those sound like good ideas. Right now, we are mostly relying on players to take care of markers (I was working under the assumption that players would not order their stunned creatures to attack anything), but there are definitely quite a few cases where we should auto-handle markers. Auto-exhausting action markers would not be difficult at all, but let's leave it for Q2 anyway.

**Comment by sIKE23** [comment]
Two items from todays game (full game) with Schwenkgott:

> [Schwenkgott] I have the critical strike enchantment on the wolf, never got the chance to reveal it. it only asks for reveal of lion savagery
> schwenkgott pays 3 mana.
>schwenkgott reveals Critical Strike!
> [sike23] ok


> [Schwenkgott] when my wolf is attacking, it asks before AND after the dice roll if i want to reveal the lion savagery
> [sike23] so no the arrow doesnt work
> [Schwenkgott] after dice roll, that makes no sense
> [Schwenkgott] and the same with the attack from the damage barrier, i keep getting asked if i want to reveal lion savagery
> [sike23] I will ask but those might be generic options but I will find out
> [Schwenkgott] but it makes no sense in a defensive action
>  [sike23] i will bring it up

**Comment by ACG8** [comment]
Regarding Critical Strike: It was overlooked. I'll add it in.

Regarding Lion Savagery timing: Lion Savagery has 2 effects; Charge +2 and Piercing +1. Charge +2 is relevant before rolling dice (to modify the number rolled). Piercing +1 is relevant after rolling dice (to boost damage once you have seen how much normal vs. critical damage is done). The choice of whether to reveal it is thus relevant before and after rolling dice. If you think I should restrict the prompt to only before rolling dice, though, I can certainly do that.

Regarding offensive enchantment prompts on defense: I am aware of this. Mostly, it was just easier to check the name of the enchantment against the phase without worrying about attacker vs. defender. I can look into modifying this, though.

Also, is the current enchantment prompt menu appropriate? I have some concerns about revealing too much information (by the amount of time taken between steps), but at the same time I don't want to prompt the players after every step; that would be annoying. If you have a suggestion for a better way to implement it, let me know.

**Comment by ACG8** [comment]
Okay, I added Critical strike. Also, I moved the charge prompt to before the attack menu. Let me know if you think this works better; I left the old code in (but commented out), so it would be easy to change back. This means, of course, that Lion Savagery will need to be revealed before declaring an attack in order to get the charge bonus. Therefore I deleted the pre-dice roll prompt, leaving only the prompt after the dice roll for Lion Savagery, in case players want some piercing.

**Comment by sIKE23** [comment]
On a solo game it is looping the attack. Had to join as a spectator to get the Chat Log....

Bloodcrag Minotaur inflicts 4 damage on Orc Butcher! (an above average roll)
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 1 normal damage, 0 critical damage, and 4 on the effect die
sIKE23 has elected not to apply auto-calculated battle results
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 1 normal damage, 2 critical damage, and 10 on the effect die
Bloodcrag Minotaur inflicts 3 damage on Orc Butcher... (a below average roll)
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 0 normal damage, 3 critical damage, and 3 on the effect die
sIKE23 has elected not to apply auto-calculated battle results
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 0 normal damage, 3 critical damage, and 3 on the effect die
sIKE23 has elected not to apply auto-calculated battle results
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 2 normal damage, 3 critical damage, and 5 on the effect die
sIKE23 has elected not to apply auto-calculated battle results
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 1 normal damage, 3 critical damage, and 10 on the effect die
Bloodcrag Minotaur inflicts 4 damage on Orc Butcher! (an above average roll)
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 4 normal damage, 3 critical damage, and 6 on the effect die
Bloodcrag Minotaur inflicts 7 damage on Orc Butcher! (an above average roll)
Bloodcrag Minotaur attacks Orc Butcher with Horns!
sIKE23 rolls 4 attack dice 
sIKE23 rolled 0 normal damage, 3 critical damage, and 1 on the effect die

I clicked yes/no/X in upper right hand corner

I will try a 2 player game a bit later tonight

**Comment by sIKE23** [comment]
Look at the chat log in testing folder on the Dropbox\MageWars-Dev\Testing\LoopingAttacks.txt

Luckily with two-players you can discard one of the two offenders to break the loop. I have been able to reproduce this a couple of times tonight. A Unicorn with Lion Savagery and a Minotaur with Lion Savagery melee attacking each other.

**Comment by ACG8** [comment]
Okay, I'll take a look. I can't imagine what the problem could be; perhaps my last change resulted in something unexpected. Probably won't be able to get to it until this evening at the earliest, though.

**Comment by ACG8** [comment]
I have identified the problem (something is interfering with the way attacks keep track of what the unmodified attack was), but it is resisting my attempts to fix it. This puzzle will take a bit more time to solve.

**Comment by ACG8** [comment]
I finally managed to fix it. The problem was that the copy of the original attack made in computeAttack was not deep enough, and was still referring to the original attack traits sub-dictionary object (rather than creating a copy of it). I don't suppose there is any chance we can import the ```copy.py``` module, is there? I really miss ```deepcopy().```

**Comment by ACG8** [comment]
Also, though I don't think my fix should affect other things, be on the lookout for unexpected behavior for attacks that make multiple strikes, just in case.

**Comment by sIKE23** [comment]
That is good to know, as I tested the Hydra the other day without incident
I will give it another round of testing here soon.

**Comment by sIKE23** [comment]
For Mages, when using either an attack spell (Sunfire Burst/Fireball/Acid Ball were tested) or Ranged Attack (Ivarium Longbow tested), the dice roll happens but the Apply Damage prompt doesn't not appear. The apply damage prompt does work for Mages melee attacks. Dazes that were rolled are applied.

Other testing notes:

Darkfenne Hydra - Triple Strike works as expected
Gorgon Archer - Works as expected, however she was able to land a Weak Marker (included in the BC pre-roll prediction) on a Skeletal Sentry which is Poison Immune.

Being able to create cards on the fly is very nice!

**Comment by ACG8** [comment]
```For Mages, when using either an attack spell (Sunfire Burst/Fireball/Acid Ball were tested) or Ranged Attack (Ivarium Longbow tested), the dice roll happens but the Apply Damage prompt doesn't not appear.```

```Gorgon Archer - Works as expected, however she was able to land a Weak Marker (included in the BC pre-roll prediction) on a Skeletal Sentry which is Poison Immune.```

Both issues should be fixed now.

**Comment by sIKE23** [comment]
Is it possible to convert the code in lines 772 - 776 in attacks.py into a dedicated creature/conjuration death check? Why? During Upkeep we process Burns/Rot/Bleed/Etc and apply damage it would be very cool to be able to notify a Death Msg if the applied damage would kill the creature.

**Comment by sIKE23** [comment]
If you have Bear Strength hidden on a creature or mage when it starts the attack sequence, you are prompted to see if you want to reveal it, if you say yes, you are prompted to pay for the enchantment, it is revealed but the melee bonus is not applied to this attack sequence. It is applied to the next one  though.

**Comment by ACG8** [comment]
> Is it possible to convert the code in lines 772 - 776 in attacks.py into a dedicated creature/conjuration death check? Why? During Upkeep we process Burns/Rot/Bleed/Etc and apply damage it would be very cool to be able to notify a Death Msg if the applied damage would kill the creature.

Yes, it is. The one (minor) complication is that the getRemainingLife function requires passing a creature dictionary, which means calling the computeTraits() function. ComputeTraits() is useful for other things as well, but calling it too many times may cause noticeable lag. The way to solve this would be to call it exactly once for each object in play, store the resulting dictionaries in a list, and then use those to resolve upkeep effects.

So, in general, what you would do during upkeep is first create a list of objects:

```objectDictList = [computeTraits(c) for c in table if c.isFaceUp]```

Then, you iterate over the list, performing all the upkeep-phase operations on its contents. For example:

```
for oDict in objectDictList:
    object = Card(oDict.get("OwnerID")) #This gives you the card object corresponding to the dictionary)
    #Upkeep functions
    upkeep = oDict.get("Upkeep")
    resolveUpkeepFunctionHere(object,upkeep)
    #etc...
```

Then, at the end (or earlier if desired), you could check the remaining life of the object and give a deathprompt if necessary:
```
    #...
    if object.Type in ["Mage","Creature","Conjuration","Conjuration-Wall"] and not oDict.get("Indestructible") and getRemainingLife(oDict) == 0: deathPrompt(oDict)
```

The deathPrompt() function is already set up so that it can handle non-attack causes of death; you just omit the attack and attackerDict arguments.

> If you have Bear Strength hidden on a creature or mage when it starts the attack sequence, you are prompted to see if you want to reveal it, if you say yes, you are prompted to pay for the enchantment, it is revealed but the melee bonus is not applied to this attack sequence. It is applied to the next one though.

I will look into this.

**Comment by sIKE23** [comment]
Schwenkgott tested with Mystery today:
 
![schwenkgotts issues](https://cloud.githubusercontent.com/assets/5695175/6993621/8ee39c66-dabf-11e4-84dd-d9211c0f9c1a.jpg)

 
![sectarus attack](https://cloud.githubusercontent.com/assets/5695175/6993619/8b2830b4-dabf-11e4-9d1b-7f2aef50281f.jpg)

**Comment by ACG8** [comment]
I find it odd that (6) worked even once, as I never coded for Agony at all (an oversight that I will fix, of course).

**Comment by ACG8** [comment]
Fixed everything except problem (4) which (as noted) is complicated. I'll open a separate issue for that.

**Comment by sIKE23** [comment]
![tainted error](https://cloud.githubusercontent.com/assets/5695175/7203283/26b092e6-e4df-11e4-96f3-676a3788031f.jpg)
I will fix

**Comment by sIKE23** [comment]
schwenkgott We did a testrun: 
schwenkgott 1) when you have prepared sandstorm and attack the enemy, you will get no attack sequence started. there is a bug with the sandstorm - fixed
schwenkgott 2) Heart of Gravikor (Melee Weapon) creates error upon attacking - fixed
schwenkgott 3) its possible to initiate an attack sequence although you are stunned. 
schwenkgott 4) Push effect from Windstorm is not working - we do not move cards...so if the roll is successful you will need to move the card
schwenkgott 5) Hurl Meteorite gives no Daze (Daze, 9.) is written in chat - fixed
schwenkgott 6a) How is it possible to attack your own mage with the mage using the new attack sequence system? -  manual roll for now
schwenkgott 6b) There are two different Windstorms in the deck builder. one has 5+ push and one has 6+ push on the picture, but both have 6+ push in the stats of the card - release version of the card has +6 
schwenkgott 6c) Geysir used on a target thats burning will be normal until now with damage and daze chance. is there a way to check for extinguish? - Geyser doesn't have the Extinguish trait...
schwenkgott 7) In the Attack Sequence i am to reveal Marked for Death, but after reveal I do not get I more dice - should work correct?
schwenkgott 8) A readied Bim Shalla is not counted for buffing the attack during attack sequence - next release
schwenkgott 9) Fumble is recognized as an enchantment that can be revealed in the attack sequence. but revealing it Will not result in a missing attack dice are rolled - should work correct?
schwenkgott 10) how are the heal spells working with the new attack sequence system? how are other incantations working? (buffs for one round only) - next release
schwenkgott 1I) Zone Exclusive is not in yet: you can place several zone exclusive conjurations in the same zone - next release (I think)
schwenkgott 12a) defenses work (tested on cevere with melee attack), but after a defense is used, that ready marker is not flipped automatically. if cevere is attacked a second time (defense marker still ready, but defense use before), cevere cannot use defense anymore. seems that only the ready marker is the problem here - can we get this in for Q1?
schwenkgott 12b) I cannot add tainted manually to a creature (alt t not working, error is written in chat), schwenkgott you got a screen in tester folder for that - i will fix
schwenkgott 13) Defense from Forcemaster is not working - should work, correct?
schwenkgott 14) plaguemaster is not working, there is no prompt for adding damage to poisoned enemies - should work, correct?
schwenkgott 15) eternal servant on legendary creatures is counted in as piercing +1 - should work, correct?
schwenkgott 16) no prompt for reanimating eternal servant upon death - next release, I think
schwenkgott 17) activating voltaric shield is not taking 2 mana automaticly - next release?
schwenkgott 18) Malakais Fire not working - should work correct?
schwenkgott 19) Moving BloodReaper to a Demon does not take away mages life - next release?
schwenkgott 19a) Same for Pet Holy Avenger (manawise) - next release?
schwenkgott 20) Harshforge plate is not increasing casting cost from enemy enchantments - next release I think?
schwenkgott 21) In the attack sequence with adramelech warlock, there is no possibility to reveal a curse on the target to buff a flame attack by I dice. tested with rust. there was a prompt for revealing rust, but AFTER the roll dice step. so i missed one dice in the attack - should we fix this?

 schwenkgott - weapons and attack spells are done testing, only missing mage only weapons

**Comment by ACG8** [comment]
> schwenkgott 3) its possible to initiate an attack sequence although you are stunned. 
I guess I could change that.

> schwenkgott 6c) Geysir used on a target thats burning will be normal until now with damage and daze chance. is there a way to check for extinguish? - Geyser doesn't have the Extinguish trait...

Extinguish is not currently implemented. Didn't think it was that important.

>schwenkgott 7) In the Attack Sequence i am to reveal Marked for Death, but after reveal I do not get I more dice - should work correct?

Will look into this.

>schwenkgott 12a) defenses work (tested on cevere with melee attack), but after a defense is used, that ready marker is not flipped automatically. if cevere is attacked a second time (defense marker still ready, but defense use before), cevere cannot use defense anymore. seems that only the ready marker is the problem here - can we get this in for Q1?

I'll take a look.

>schwenkgott 12b) I cannot add tainted manually to a creature (alt t not working, error is written in chat), schwenkgott you got a screen in tester folder for that - i will fix

Sorry about that; I fixed it as soon as I saw the error message image. Hope that didn't cause too much sync trouble.

>schwenkgott 13) Defense from Forcemaster is not working - should work, correct?
Haven't looked at it before. Probably missing from the XML definition.

>schwenkgott 14) plaguemaster is not working, there is no prompt for adding damage to poisoned enemies - should work, correct?

This was never implemented.

>schwenkgott 15) eternal servant on legendary creatures is counted in as piercing +1 - should work, correct?
Yes, this one should be working. Let me know if it isn't (well, I'll test it anyway when I have time).

>schwenkgott 16) no prompt for reanimating eternal servant upon death - next release, I think
schwenkgott 17) activating voltaric shield is not taking 2 mana automaticly - next release?
schwenkgott 18) Malakais Fire not working - should work correct?
schwenkgott 19) Moving BloodReaper to a Demon does not take away mages life - next release?
schwenkgott 19a) Same for Pet Holy Avenger (manawise) - next release?
schwenkgott 20) Harshforge plate is not increasing casting cost from enemy enchantments - next release I think?

Those should all be left for the next release. I suppose I could look into Malakai's Fire, but it seems like more trouble than it is worth.

>schwenkgott 21) In the attack sequence with adramelech warlock, there is no possibility to reveal a curse on the target to buff a flame attack by I dice. tested with rust. there was a prompt for revealing rust, but AFTER the roll dice step. so i missed one dice in the attack - should we fix this?

That is because the script that decides when enchantments should be revealed does not take into account anything besides the effect of the enchantment itself. It will be challenging to fix this.

I don't have time right now, but I'll look at these a bit later. Are we still planning to release today (and then just issue a patch), or do you want to wait until the fixable ones are fixed?

**Comment by sIKE23** [comment]
If you get the time to fix these issues that would be great...I can delay
by a day or two no problem, unless you want it out there....

**Comment by ACG8** [comment]
I have some time. Let me see how many I can get through.

**Comment by ACG8** [comment]
Okay, I just fixed the problems that were relatively simple. I tested them, but maybe it would be a good idea for you to have them tested anyway. How do you feel about holding off on the rest of schwenkgott's un-fixed concerns? Are there any remaining that you think are essential for this release?

**Comment by sIKE23** [comment]
You left the shortcuts for Sleep the same as for Slam and there are no more
pairs of S's available, I think Ctrl+P and Ctrl+Shift+P are the best
candidates. I will fix the shortcut and roll another beta build and let
Schwenkgott know....

</sIKE>

**Comment by sIKE23** [comment]
When Fumble is revealed by the BC and the Casting Costs are paid, the Attack should be canceled.

**Comment by ACG8** [comment]
This appears to be working as intended; I opened up a two player game with warlock v. priest, enchanted a cleric with fumble, and had it attack a dark pact slayer. Upon revealing fumble, I got the message "Asyran Cleric fumbles the attack!"

![image](https://cloud.githubusercontent.com/assets/10094820/7524456/aff6e1f2-f4c7-11e4-8927-1fba9a03edb8.png)

In what circumstances is this not working?

**Comment by sIKE23** [comment]
That is not what we were seeing, Try placing the Fumble on the Mage and
then have him make an attack....

**Comment by ACG8** [comment]
Oh, okay. I'll take another look. If that is the problem, I think I can already guess what is causing it.

**Comment by ACG8** [comment]
No, it is still working for me...

![image](https://cloud.githubusercontent.com/assets/10094820/7525301/ca6ae442-f4cd-11e4-87f4-6825bc0cfefa.png)

**Comment by sIKE23** [comment]
Both players should be able to attach the same hidden enchantment on to the appropriate object. Think Rise Again...

**Comment by sIKE23** [comment]
The BC does not handle the Wizard's Tower correctly. 

Test: Prepare it, cast and pay for it, Bind legal hidden attack spell to the WT (the spell does bind), use the WT to launch a legal attack, you are prompted to pay the casting cost of the WT and the BC doesn't activate.

**Comment by sIKE23** [comment]
Does the BC handle the Dwarf Warlords runes?

**Comment by ACG8** [comment]
>Does the BC handle the Dwarf Warlords runes?

Yes; at least, it should handle the Armor +1, Defense +2, and Piercing +1 runes (but not the discount or cantrip runes)

**Comment by sIKE23** [comment]
Tried to bind a Stranglevine on to a Vine Tree and got the following error message:

![vinetreebug](https://cloud.githubusercontent.com/assets/5695175/7551556/63e84a4e-f652-11e4-8746-4378a47e6bd0.jpg)

**Comment by ACG8** [comment]
>Both players should be able to attach the same hidden enchantment on to the appropriate object. Think Rise Again...

I just tested this, and there doesn't seem to be any difficulty with it (I ran two instances of OCTGN, and was able to have each player attach a face-down Rise Again to the same Bog Queen Acolyte). In what conditions does this not work?

**Comment by sIKE23** [comment]
I will find out......what card was having the issue. Did you see the one
with Decoy?

**Comment by ACG8** [comment]
Yes, see cbdf0ea6e305d211201b2aa0d74de6db0e1af7d3

**Comment by sIKE23** [comment]
Schwenkgott is reporting that the Joktari BM's Ranged +1 is being applied to Attack Spells when it shouldn't.

**Comment by ACG8** [comment]
On it.

**Comment by sIKE23** [comment]
Weird issue - Reflex Boots are on an attacking Mage, the defending mage gets the prompt to use the defense of the Reflex Boots.

**Comment by sIKE23** [comment]
Lion Savagery - has [Charge +2] but not [Piercing +1] are you handling that in the code or do I need to add [Piercing +1] to the set data?

**Comment by ACG8** [comment]
Lion Savagery's Piercing +1 only applies to melee attacks, so I had to handle it in the code. You don't need to add it to the set data.

**Comment by sIKE23** [comment]
The JBM's Wounded Prey Marker is moved to the opposing Mage when a creature he (the opposing mage) controls is destroyed.

**Comment by sIKE23** [comment]
Ranged Attacks removes the Guard Marker from a target

**Comment by sIKE23** [comment]
Restrained or Incapacitated flying creatures cannot be targeted by ground units, I don't think when you extend these traits that the Flying trait is either removed or ignored.

**Comment by ACG8** [comment]
I'll get right on those (though not until this evening). I am also making progress on the Domination Map loading, though there are still a few bugs to work out.

**Comment by sIKE23** [comment]
Fantastic news on the Domination side.

I am getting better at understanding your code (I think) here is my stab at
fixing this flying issue (starting at line 133 of attacks.py)

        if (dTraitDict.get('Flying') and
            not (dTraitDict.get('Restrained') or
            not (dTraitDict.get('Incapacitated') and
            not aTraitDict.get('Flying') and
            attack.get('RangeType') == 'Melee' and
            not atkTraits.get('Reach')): return False

I think adding an indent to line 862 will fix the guard issue...

Don't tell me until you fix it and I will look at your code on Github...

**Comment by sIKE23** [comment]
Unrevealed Damage Barriers are making attacks.

**Comment by ACG8** [comment]
>        if (dTraitDict.get('Flying') and
            not (dTraitDict.get('Restrained') or
            not (dTraitDict.get('Incapacitated') and
            not aTraitDict.get('Flying') and
            attack.get('RangeType') == 'Melee' and
            not atkTraits.get('Reach')): return False

Yes, that would certainly work. However, the root cause of the issue is that the computeTraits() function was allowing restrained/incapacitated flying creatures to retain the flying trait, so I just fixed it so that it does not.

>I think adding an indent to line 862 will fix the guard issue...

It does. Nice.

**Comment by sIKE23** [comment]
An attached Tanglevine on a flyer (say Slammed/Guard/Stunned) should be
able to be attacked by a non-flyers melee attack

**Comment by sIKE23** [comment]
I am trying to understand extend, append, and remove. Any helpful points?

//FC

**Comment by ACG8** [comment]
>An attached Tanglevine on a flyer (say Slammed/Guard/Stunned) should be
able to be attacked by a non-flyers melee attack

Hmm... odd that it can't be. I'll check it out.

>I am trying to understand extend, append, and remove. Any helpful points?

Sure. I assume you are talking about the use of these in the computeTraits() function.

The issue that I was trying to solve was that computeTraits is both heavily used and rather lengthy, so it is very important to make sure it takes as little time as possible. Retrieving functions takes time. Therefore, in lines 1148-1150, I defined functions append, extend, and remove:

```
append = rawTraitsList.append
extend = rawTraitsList.extend
remove = rawTraitsList.remove
```

This way, Python only has to look up these functions once, to define functions append, extend, and remove. After that, I can call these three functions without having to look them up, which saves time.

For example, in line 1267, I could write either

```if markers[Melee]: append('Melee +{}'.format(str(markers[Melee])))```

or 

```if markers[Melee]: rawTraitsList.append('Melee +{}'.format(str(markers[Melee])))```

and they would do the same thing. However, the first statement is faster, because python already knows what append() does (because I defined it earlier in the computeTraits function), whereas it has to look up what rawTraitsList.append() does. By looking up these functions only once, I can shave a lot of time off of computeTraits(), which is essential to reduce noticeable lag.

**Comment by ACG8** [comment]
>An attached Tanglevine on a flyer (say Slammed/Guard/Stunned) should be
able to be attacked by a non-flyers melee attack

I was not able to reproduce this issue. What were the conditions under which this occurred?

![image](https://cloud.githubusercontent.com/assets/10094820/7785650/0088cccc-015f-11e5-8df5-3cd0402a46ab.png)

**Comment by sIKE23** [comment]
I tested the following way:

Mage and Blood Demon same zone. Mage has no Ranged Attacks.
- Arrow from Mage to BD, No Legal Attacks detected.
- Cancel Attack
- Add Slam Marker to BD
- Arrow from Mage to BD, Basic Melee Attack
- Cancel Attack
- Cast and Attach Tanglevine to BD
- Arrow from Mage to BD, Basic Melee Attack
- Cancel Attack
- Remove Slam Marker
- Arrow from Mage to BD, No Legal Attacks detected.

What is working, you cant attach Tanglevine to BD when he is flying, once
you have him Slammed, you can attach the Tanglevine to the BD. When you
attach this Conjuration you should be prompted to "Cast" it though.

I am guessing that the Slam/Stun/Stuck/etc is working as expected (removes
the Flying trait), however when the marker is removed after the Tanglevine
is attached the BC is not keeping the removal of the Flying trait.

Thanks!

//FC

**Comment by sIKE23** [comment]
> append = rawTraitsList.append
> extend = rawTraitsList.extend

When would one choose append over extend, I am unsure of their differences...

**Comment by ACG8** [comment]
>When would one choose append over extend, I am unsure of their differences...

Append is used to add a single element to the end of a list. Extend is used to add a list of elements to the end of the list. If you are adding just 1 element, it is faster to do ```append(x)```, as opposed to ```extend([x])```. If you are adding multiple elements, it is faster to do ```extend([x,y,z])``` rather than appending each element separately.

**Comment by ACG8** [comment]
>Mage and Blood Demon same zone. Mage has no Ranged Attacks.
- Arrow from Mage to BD, No Legal Attacks detected.
- Cancel Attack
- Add Slam Marker to BD
- Arrow from Mage to BD, Basic Melee Attack
- Cancel Attack
- Cast and Attach Tanglevine to BD
- Arrow from Mage to BD, Basic Melee Attack
- Cancel Attack
- Remove Slam Marker
- Arrow from Mage to BD, No Legal Attacks detected.

I tried this, but I am not getting the error. I only fixed the tanglevine issue in the latest update, though; have you tried with the latest version?

Good point about prompting to cast the tanglevine.

**Comment by sIKE23** [comment]
Should be the same for both Tangle and Strangle vines.

I just dropped a video on the temp folder on the dropbox dev share

**Comment by sIKE23** [comment]
Did you ever get a chance to look at this? If you like I can schedule a
time with you to get on a game and reproduce the issue....

</sIKE>

**Comment by sIKE23** [comment]
Revealing Forcefield does not cancel the Attack.

**Comment by sIKE23** [comment]
Dancing Scimitar's attack benefits from Bear Strength and from other melee buffs like the Gauntlets of Strength.

**Comment by ACG8** [comment]
>Did you ever get a chance to look at this? If you like I can schedule a
time with you to get on a game and reproduce the issue....

Sorry; my workload has really increased lately, and I haven't had much time to work on the module. I'll take a look at these issues this weekend.

**Comment by sIKE23** [comment]
I understand, it was a rough travel week for me. I just got home myself,
the weather is making travel difficult! I made an update and pushed it up
to github today...

**Comment by sIKE23** [comment]
Arcane Zap has two issues, it should notify that the Mage pays 1 mana when cast. I have fixed that, the other is that the game should have memory of Zap being used as it is allowed to be used only once a round.

**Comment by sIKE23** [comment]
Create a Vine Marker, then prepare a Burst of Thorns, now use an Arrow from the Burst of Thorns to the Vine Marker:

> Traceback (most recent call last):
> File "D:\Users\fczajka\Documents\Octgn\GameDatabase\9acef3d0-efa8-4d3f-a10c-54812baecdda\scripts\actions.py", line 1860, in castSpell
> File "D:\Users\fczajka\Documents\Octgn\GameDatabase\9acef3d0-efa8-4d3f-a10c-54812baecdda\scripts\actions.py", line 2036, in computeCastCost
> File "<string>", line 1, in <module>
> File "D:\Users\fczajka\Documents\Octgn\GameDatabase\9acef3d0-efa8-4d3f-a10c-54812baecdda\scripts\actions.py", line 314, in onTargetCardArrow
> File "D:\Users\fczajka\Documents\Octgn\GameDatabase\9acef3d0-efa8-4d3f-a10c-54812baecdda\scripts\actions.py", line 2036, in <lambda$28>
> ValueError: Invalid integer literal
> onTargetCardArrow(Player(1), Card(65596), Card(65625), True)

**Comment by ACG8** [comment]
>Should be the same for both Tangle and Strangle vines.

>I just dropped a video on the temp folder on the dropbox dev share

Found the issue. It turns out it was due to the duplicate tanglevine xml code from DvN. I've been meaning to ask about this for a while; cards that appear in multiple sets are inconvenient, because we have to make sure all the xml code is correct for each one whenever we change one.

However, the SpellDictionary will render this moot (as with many other things), since the code relevant parts will be contained in the python code, rather than the xml code.

**Comment by sIKE23** [comment]
Damage Barriers are still being triggered even if the Attack was Avoided (Defense, failed Daze roll, Fumble) it should trigger after a successful melee attack i.e Attack Dice are rolled...

**Comment by ACG8** [comment]
Are there any issues on 144 for the master branch that still need to be resolved?

**Comment by sIKE23** [comment]
I have nothing at this point, I have been keeping it open, as most of the
issues that bubble up involve it.

Did you like the cardSizes work and style of coding I did today?

Thanks!

</sIKE>

**Comment by ACG8** [comment]
Yes, looks good! I like the neat indexing of the sizes in the dictionary.

Going to see about that Domination code now...

**Comment by sIKE23** [comment]
Wildfire Imps are not getting their Melee +2 buff when they are attack a Burning Object.

**Comment by sIKE23** [comment]
The Bloodfire Helmets Melee +1 buff for Demons controlled by the Warlock.

```python
if attacker.Subtype == "Demon" and attacker.controller == ?? and defender.markers[Burn]:
 localADict['Melee'] = localADict.get('Melee',0) + 1
```
For the ?? I don't know how to form the attacker.controller "has the Bloodfire Helmet" equipped ....

**Comment by ACG8** [comment]
I'll fix the bloodfire helm problem (tomorrow, most likely). It sounds like you have a handle on the Wildfire Imp issue, so I'll leave that to you.

**Comment by sIKE23** [comment]
Armor +1 Markers are not calculated by the BC.

**Comment by sIKE23** [comment]
The results of a successful Vampiric attack is handled and calculated properly. However the wording used does not make sense at all. There is  no "draining" like this says:

> Blood Demon drains 2 health from Bridge Troll!

I am trying to come up with better phrasing but it is late at night and I will stew on it for a while. If I come up with something better, would you mind me changing it after a review by you?

**Comment by sIKE23** [comment]
Ignore my Armor +1 Marker comment don't know why I was thinking that it wasn't working.

> {} Vampiricly heals {} of the damage inflicted on {}!

and it would look like:
> Blood Demon Vampiricly heals 2 from the damage inflicted on Bridge Troll!

and the question would be:

> Remove {} damage health from {} for Vampiric attack on {}?

and it would look like:
> Remove 2 damage health from Blood Demon for Vampiric attack on Bridge Troll?

**Comment by ACG8** [comment]
I see no problem with the current wording, but perhaps:

> {} absorbs {} health from {}!

or

>{} heals {} damage through vampirism!

What is your objection to "drain?"  That verb is commonly used in many games for vampiric mechanics.

**Comment by sIKE23** [comment]
It sounds like it is draining more life away from the defender..

**Comment by ACG8** [comment]
I see; that makes sense. I think the wording I would prefer (for simplicity and brevity) then would be

>{} heals {} damage through vampirism!

and

>Heal {} damage through vampirism?

**Comment by sIKE23** [comment]
Now if we want to add in flavor of the Vampiric Draining we can add that in
in events file during his attack.(?)
